idea/src/hev-timeline/hev-interpolateTranslation.1
.TH hev-interpolateTranslation 1 "November 2011"

.SH NAME
hev-interpolateTranslation

.SH SYNOPSIS 
.B hev-interpolateTranslation  [ --t0 \fIx y z\fB ] [ --t1 \fIx y z\fB ]  [ --absolute ] [ --startTime \fIt\fB ] [ --duration \fIt\fB ] [ --frameRate \fIr\fB ] [ --deltaTranslation \fIt\fB ]

.SH OPTIONS

.IP "--t0 \fIx y z\fR"
specifies the initial position. If not specified, 0, 0, 0, is used.

.IP "--t1 \fIx y z\fR"
specifies the final position. If not specified, 0, 0, 0, is used.

.IP "--absolute"
specifies that the times used in the IRIS control commands be absolute
times. i.e. OSG reference times. The default is to use time values that are
relative to the time the control command was processed.

.IP "--startTime \fIt\fR"
specifies the initial time offset in seconds to be added to all IRIS control
commands.  The default value is 0.

.IP "--duration \fIt\fR"
specifies the number of seconds to move from t0 to t1. If not supplied,
duration will be set to 1.0 second.

.IP "--frameRate \fIr\fR"
specifies the number of times per second \fBhev-interpolateTranslation\fR
will interpolate between the positions. If \fIframeRate\fR is not specified, 60 will be used.

.IP "--deltaTranslation \fIt\fR"
specifies the amount that the navigation node's translation value needs to
change before a command is written. The amount is the distance from the
last written translation.

If \fIdeltaTranslation\fR is not specified, it defaults to .001.

If \fIdeltaTranslation\fR is set to zero a command containing a translation
value will be written \fIframeRate\fR times a second.

.SH DESCRIPTION

\fBhev-interpolateTranslation\fR will write IRIS control commands that will
modify the navigation node's translation to go from \fIt0\fR to
\fIt1\fR over the course of \fIduration\fR seconds, writing up to \fIframerate\fR commands
per second.

\fBhev-interpolateTranslation\fR liberally comments the control commands it
generates. Like any IRIS control command file, the results can be edited with any text
editor.

.SH EXAMPLE

hev-interpolateTranslation --t0 1 1 1 --t1 10 20 30 --start 10 --duration 2.5 --frameRate 4 > /tmp/scoot.iris

will generate and store in the file /tmp/scoot.iris:

AFTER TIME 10.000000 NAV POSITION 1.000000 1.000000 1.000000
.br
AFTER TIME 10.250000 NAV POSITION 1.900000 2.900000 3.900000
.br
AFTER TIME 10.500000 NAV POSITION 2.800000 4.800000 6.800000
.br
AFTER TIME 10.750000 NAV POSITION 3.700000 6.700000 9.700001
.br
AFTER TIME 11.000000 NAV POSITION 4.600000 8.600000 12.600000
.br
AFTER TIME 11.250000 NAV POSITION 5.500000 10.500000 15.500000
.br
AFTER TIME 11.500000 NAV POSITION 6.400000 12.400001 18.400002
.br
AFTER TIME 11.750000 NAV POSITION 7.300000 14.300000 21.299999
.br
AFTER TIME 12.000000 NAV POSITION 8.200001 16.200001 24.200001
.br
AFTER TIME 12.250000 NAV POSITION 9.099999 18.100000 27.099998
.br
AFTER TIME 12.500000 NAV POSITION 10.000000 20.000000 30.000000



.SH SEE ALSO

hev-timeline, hev-navigationToInterpolate, hev-timelineToIris, hev-interpolateRotation,
hev-interpolateScale, hev-readNav

.PP
John Kelso, kelso@nist.gov
.br
NIST High Performance Computing and Visualization Group
.br
November 2011
.br
\fBhttp://www.nist.gov/mcsd/hpcvg/\fR



idea/src/hev-timeline/hev-interpolateTranslation.cpp
#include <stdio.h>
#include <iostream>
#include <stdlib.h>

#include <iris.h>

#include "utils.h"

void usage()
{
    fprintf(stderr,"Usage: hev-interpolateTranslation [ --t0 x y z ] [ --t1 x y z ] [ --absolute ] [ --startTime t ] [ --duration t ] [ --frameRate r ] [ --deltaTranslation t ]\n") ;
}

////////////////////////////////////////////////////////////////////////
int main(int argc, char** argv)
{

    // send dtk messages to stdout
    dtkMsg.setFile(stderr) ;

    osg::Vec3d t0(0,0,0) ;
    osg::Vec3d t1(0,0,0) ;
    float startTime = 0 ;
    float duration = 1 ;
    float frameRate = -1 ;
    bool absolute = false ;
    double deltaTranslation = .001 ;

    int c = 1 ;
    while (c < argc && argv[c][0] == '-' && argv[c][1] == '-')
    {
	if (c>=argc || iris::IsSubstring("--absolute",argv[c],3))
	{
	    c++ ;
	    absolute = true ;
	}
	else if (c>=argc || iris::IsSubstring("--startTime",argv[c],4))
	{
	    c++ ;
	    if (c>=argc || !iris::StringToFloat(argv[c],&startTime) || startTime<0) 
	    {
		dtkMsg.add(DTKMSG_ERROR, "hev-interpolateTranslation: invalid startTime value\n") ;
		usage() ;
		return 1 ; 
	    }
	    c++ ;
	}
	else if (c>=argc || iris::IsSubstring("--duration",argv[c],4))
	{
	    c++ ;
	    if (c>=argc || !iris::StringToFloat(argv[c],&duration) || duration<=0) 
	    {
		dtkMsg.add(DTKMSG_ERROR, "hev-interpolateTranslation: invalid duration value\n") ;
		usage() ;
		return 1 ; 
	    }
	    c++ ;
	}
	else if (c>=argc || iris::IsSubstring("--frameRate",argv[c],4))
	{
	    c++ ;
	    if (c>=argc || !iris::StringToFloat(argv[c],&frameRate) || frameRate<=0) 
	    {
		dtkMsg.add(DTKMSG_ERROR, "hev-interpolateTranslation: invalid frameRate value\n") ;
		usage() ;
		return 1 ; 
	    }
	    c++ ;
	}
	else if (c>=argc || iris::IsSubstring("--deltaTranslation",argv[c],3))
	{
	    c++ ;
	    if (c>=argc || !iris::StringToDouble(argv[c],&deltaTranslation) || deltaTranslation<0 ) 
	    {
		dtkMsg.add(DTKMSG_ERROR, "hev-interpolateTranslation: invalid deltaTranslation value\n") ;
		usage() ;
		return 1 ; 
	    }
	    c++ ;
	}
	else if (c>=argc || !strcmp("--t0",argv[c]))
	{
	    c++ ;
	    if (c>=argc-2 || !iris::StringToDouble(argv[c],&(t0.x())) || !iris::StringToDouble(argv[c+1],&(t0.y())) || !iris::StringToDouble(argv[c+2],&(t0.z())))
	    {
		dtkMsg.add(DTKMSG_ERROR, "hev-interpolateTranslation: invalid --t0 value\n") ;
		usage() ;
		return 1 ; 
	    }
	    c+=3 ;
	}
	else if (c>=argc || !strcmp("--t1",argv[c]))
	{
	    c++ ;
	    if (c>=argc-2 || !iris::StringToDouble(argv[c],&(t1.x())) || !iris::StringToDouble(argv[c+1],&(t1.y())) || !iris::StringToDouble(argv[c+2],&(t1.z())))
	    {
		dtkMsg.add(DTKMSG_ERROR, "hev-interpolateTranslation: invalid --t1 value\n") ;
		usage() ;
		return 1 ; 
	    }
	    c+=3 ;
	}
	else 
	{
	    dtkMsg.add(DTKMSG_ERROR, "hev-hev-interpolateTranslation: unrecognized option: %s\n",argv[c]) ;
	    usage() ;
	    return 1 ;
	}
    }

#if 0
    // get the frame rate from the system if frameRate not passed as a parameter
    if (frameRate < 0) frameRate = getDefaultFrameRate() ;
#else
    if (frameRate < 0) frameRate = 60 ;
#endif

    std::string timeString ;
    if (absolute) 
    {
	timeString = "ABSTIME" ;
	printf("# time is relative to start of first graphics frame\n") ;
    }
    else 
    {
	timeString = "TIME" ;
	printf("# time is relative to when control commands are received\n") ;
    }
    
    printf("# moving from t0 = (%f %f %f) to t1 = (%f %f %f) ; start time = %f, duration = %f, frame rate = %f\n",t0.x(), t0.y(), t0.z(), t1.x(), t1.y(), t1.z(), startTime, duration, frameRate) ;

    Posn p0, p1;
    p0.pos = t0 ;
    p0.t = startTime ;
    p1.pos = t1 ;
    p1.t = startTime + duration ;

    Interpolate interp(p0, p1) ;

    // always write the first
    printf("# first t = %f\n",p0.t) ;
    printf("AFTER %s %f NAV POSITION %f %f %f\n",timeString.c_str(), p0.t, p0.pos.x(), p0.pos.y(), p0.pos.z()) ;
    Posn lastPosn = p0 ;

    for (int i=1; i<frameRate*duration; i++)
    {
	double t = p0.t + double(i)/frameRate ;
	Posn interpPosn ;
	printf("# interpolated t = %f\n",t) ;
	if (!interp.at(t, &interpPosn))
	{
	    dtkMsg.add(DTKMSG_ERROR, "hev-interpolateTranslation: interpolation error at t=%f\n",t) ;
	}
	else
	{
	    // only write if it changed from last time
	    if (movedEnoughTrans(interpPosn.pos, lastPosn.pos, deltaTranslation))
	    {
		lastPosn.pos = interpPosn.pos ;
		printf("AFTER %s %f NAV POSITION %f %f %f\n",timeString.c_str(), t, interpPosn.pos.x(), interpPosn.pos.y(), interpPosn.pos.z()) ;
		fflush(stdout) ;
	    }
	}
    }
    // always write the last
    printf("# last t = %f\n",p1.t) ;
    printf("AFTER %s %f NAV POSITION %f %f %f\n",timeString.c_str(), p1.t, p1.pos.x(),  p1.pos.y(),  p1.pos.z()) ;
    return 0 ;
}
